### PR TITLE
#220 - version to provides

### PR DESCRIPTION
**Pull request by olenagerasimova**
For #220 added version to `provides` tag in primary xml and corrected the test. Also added todo to continue.

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#222](https://github.com/artipie/rpm-adapter/issues/222) is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/92b3087c-aa65-420c-b19f-77c5be788f04); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/java/com/artipie/rpm/meta/XmlPrimary.java [hunk 1]
```diff
@@ -438,15 +431,31 @@ public Format headerRange(final int start, final int end) throws XMLStreamExcept
...
          * @return Self
          * @throws XMLStreamException On XML error
+         * @todo #220:30min We need to experimentally check that names and versions do not
+         *  mix up in `provides`: download oxygen-gtk2-1.3.4-3.el7.x86_64.rpm from test bundle100
+         *  (see TestBundle.class), add it to resources folder and create xml primary from it.
+         *  This rpm has not full `provides` information:
+         *  <rpm:entry name="liboxygen-gtk.so()(64bit)"/>
+         *  <rpm:entry name="oxygen-gtk2" flags="EQ" epoch="0" ver="1.3.4" rel="3.el7"/>
+         *  <rpm:entry name="oxygen-gtk2(x86-64)" flags="EQ" epoch="0" ver="1.3.4" rel="3.el7"/>
+         *  Make sure we write it the same way. Also we need
```
Comment (paulodamaso) [inline src/main/java/com/artipie/rpm/meta/XmlPrimary.java:445]: @olenagerasimova seems like something is missing at the end of the puzzle, please take a look

File: src/main/java/com/artipie/rpm/meta/XmlPrimary.java [hunk 2]
```diff
@@ -438,15 +431,31 @@ public Format headerRange(final int start, final int end) throws XMLStreamExcept
 
         /**
          * Add list of providers.
-         * @param providers Provider names
+         * @param name Provides names
```
Comment (paulodamaso) [inline src/main/java/com/artipie/rpm/meta/XmlPrimary.java:434]: @olenagerasimova This documentation should be fixed: the tag name is `provides`  , not `providers`, so this is not a list of providers, but a list of libraries provided by the package being represented by this `primary.xml` file.

File: src/main/java/com/artipie/rpm/meta/XmlPrimary.java (hunk 1 above)
Comment (paulodamaso) [inline src/main/java/com/artipie/rpm/meta/XmlPrimary.java:445]: @olenagerasimova The puzzle is incomplete, the sentence is unfinished here.

File: src/main/java/com/artipie/rpm/meta/XmlPrimary.java [hunk 3]
```diff
@@ -438,15 +431,31 @@ public Format headerRange(final int start, final int end) throws XMLStreamExcept
...
          * @throws XMLStreamException On XML error
+         * @todo #220:30min We need to experimentally check that names and versions do not
+         *  mix up in `provides`: download oxygen-gtk2-1.3.4-3.el7.x86_64.rpm from test bundle100
+         *  (see TestBundle.class), add it to resources folder and create xml primary from it.
+         *  This rpm has not full `provides` information:
+         *  <rpm:entry name="liboxygen-gtk.so()(64bit)"/>
+         *  <rpm:entry name="oxygen-gtk2" flags="EQ" epoch="0" ver="1.3.4" rel="3.el7"/>
+         *  <rpm:entry name="oxygen-gtk2(x86-64)" flags="EQ" epoch="0" ver="1.3.4" rel="3.el7"/>
+         *  Make sure we write it the same way. Also we need
+         * @todo #220:30min Providers entry also can have flags, `epoch` and `rel` attributes.
```
Comment (paulodamaso) [inline src/main/java/com/artipie/rpm/meta/XmlPrimary.java:446]: @olenagerasimova PLease take a look at the comment above: we don't have `providers` entries in `primary.xml`, just `provides`

File: src/main/java/com/artipie/rpm/meta/XmlPrimary.java [hunk 4]
```diff
@@ -438,15 +431,31 @@ public Format headerRange(final int start, final int end) throws XMLStreamExcept
...
+         *  Find a way to obtain this information from rpm and add it here. Do not forget about
+         *  test.
          */
-        public Format providers(final Iterable<String> providers) throws XMLStreamException {
+        public Format providers(final List<String> name, final List<String> versions)
```
Comment (paulodamaso) [inline src/main/java/com/artipie/rpm/meta/XmlPrimary.java:450]: @olenagerasimova As stated above, this method name is wrong: it should be called `provides`

File: src/main/java/com/artipie/rpm/pkg/HeaderTags.java [hunk 5]
```diff
@@ -195,10 +195,18 @@ public String sourceRmp() {
      * Get the provide name header.
      * @return Value of header tag PROVIDENAME.
      */
-    public List<String> providers() {
+    public List<String> providerName() {
```
Comment (paulodamaso) [inline src/main/java/com/artipie/rpm/pkg/HeaderTags.java:198]: @olenagerasimova Both methods have wrong names, please see above. Also, `PROVIDENAME` and `PROVIDEVERSION` constant could be changed to reflect better what we are addressing, something like `PROVIDESNAME` and `PROVIDESVERSION`.

**Review by paulodamaso — CHANGES_REQUESTED** [review]
@olenagerasimova There are some comments, most about  `provides`  field name confusion made in first implementation.

File: src/main/java/com/artipie/rpm/pkg/HeaderTags.java (hunk 5 above)
Comment (olenagerasimova) [reply src/main/java/com/artipie/rpm/pkg/HeaderTags.java:198]: @paulodamaso the enum is from `org.redline_rpm` package

**Comment by olenagerasimova** [comment]
@paulodamaso thanks, corrected. Nothing was missed in the puzzle, at first I intended to create one puzzle, but then decided to separate it into two.

**Comment by olenagerasimova** [comment]
@paulodamaso ping

**Review by paulodamaso — DISMISSED** [review]
@olenagerasimova That'll do, thank you!

**Comment by paulodamaso** [comment]
@g4s8 Please merge

**Comment by 0crat** [comment]
@paulodamaso[/z](https://www.0crat.com/u/paulodamaso) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Review by g4s8 — APPROVED** [review]

**Comment by 0crat** [comment]
Code review was too long (8 days), architects (@g4s8, @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova)) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

**Pull request merged**